Framework/PythonInterface/plugins/algorithms/fractional_indexing.py

```diff
@@ -0,0 +1,302 @@
+import itertools
+import numpy as np
+import scipy.cluster.hierarchy as hcluster
+from scipy.cluster.vq import kmeans2
+from scipy.spatial import KDTree
+
+_MAX_REFLECTIONS = 10
+
+
+def find_bases(qs, tolerance):
+    """Find bases for the higher dimensional indexing scheme
+
+    This will attempt to find a set of basis vectors for the
+    higher dimensional indexing scheme.
+
+    :param qs: list of q vectors in HKL space defining satellite peak offsets
+    :param tolerance: the tolerance how close each vector must be to be indexed
+                      using the generated scheme
+    :return: tuple of (number of dimensions, nx3 matrix of basis vectors)
+    """
+    qs = list(sort_vectors_by_norm(qs))
+    regenerate_grid = False
```

Comment:
Is this required? We do `regenerate_grid = not is_indexed(q, refs, tolerance)` which will overwrite this everytime